src/main/java/co/cask/hydrator/salesforce/SalesforceUtils.java

```diff
@@ -0,0 +1,171 @@
+   * Example for "SELECT Id FROM Opportunity" it returns "Opportunity".
+   * For nested queries returns top sObject
+   *
+   * @param query SOQL query
+   *
+   * @return sObject name
+   */
+  public static String getsObjectFromQuery(String query) {
+    SOQLParser.StatementContext statementContext = getStatementContext(query);
+    return statementContext.objectList().getText().toLowerCase();
+  }
+
+  /**
+   * Get field names from SOQL query
+   * Example for "SELECT Id, Name FROM Opportunity" it returns ["Id", "Name"].
+   *
+   * @param query SOQL query
+   * @return list of field names
+   */
+  public static List<String> getFieldsFromQuery(String query) {
+    SOQLParser.StatementContext statementContext = getStatementContext(query);
+    String fields = statementContext.fieldList(0).getText();
+
+    List<String> fieldsList = Arrays.asList(fields.split(","));
+    fieldsList.forEach(String::trim);
+
+    return fieldsList;
+  }
+
+  /**
+   * Returns Schema in cdap format from SOQL query. This method queries Salesforce SOAP api to
+   * get fields and their data types.
+   *
+   * @param credentials credentials info
+   * @param query SOQL query
+   *
+   * @return Schema for cdap
+   *
+   * @throws ConnectionException
+   */
+  public static Schema getSchemaFromQuery(AuthenticatorCredentials credentials, String query)
+    throws ConnectionException {
+
+    String sObjectName = getsObjectFromQuery(query);
+    Map<String, Field> nameToField = getAllsObjectFieldsMap(credentials, sObjectName);
+
+    List<String> queryFields = SalesforceUtils.getFieldsFromQuery(query);
+    return getSchemaWithFields(queryFields, nameToField);
+  }
+
+  private static Map<String, Field> getAllsObjectFieldsMap(AuthenticatorCredentials credentials, String sObjectName)
+    throws ConnectionException {
+    ConnectorConfig connectorConfig =
+      Authenticator.createConnectorConfig(credentials, SalesforceUtils.SOAP_API_VERSION);
+
+    PartnerConnection partnerConnection = new PartnerConnection(connectorConfig);
+    DescribeSObjectResult describeSObjectResult = partnerConnection.describeSObject(sObjectName);
+
+    Map<String, Field> nameToField = new HashMap<>();
+    for (Field field:describeSObjectResult.getFields()) {
+      nameToField.put(field.getName(), field);
+    }
+    return nameToField;
+  }
+
+  static Schema getSchemaWithFields(List<String> queryFields, Map<String, Field> nameToField) {
+
+    List<Schema.Field> schemaFields = new ArrayList<>(queryFields.size());
+    for (String fieldName:queryFields) {
+      Field salesforceField = nameToField.get(fieldName);
+      Schema.Type schemaFieldType;
+
+      // If this is relationship query (join equivalent for soql)
+      // or this is aggregate function, or nested query as field,
+      // don't convert it from string to anything as this is very error-prone.
+      if (fieldName.contains("(") || fieldName.contains(".")) {
+        schemaFieldType = Schema.Type.STRING;
```

Comment:
Format is not assumed. So fields will not be converted from string to anything.

Reply:
So if one of the fields is complex, we just return its results in String?

Reply:
what do you mean by complex fields?

Reply:
Fields that describes the relationship (https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_relationships_understanding.htm#sforce_api_calls_soql_relationships_understanding).

Reply:
This is needed just so the failure does not happen during schema generation. The failure will happen anyway since Bulk API won't support relationship queries and aggregate functions. But since we have a check here, user will get a more beautiful message from the API. "... is not support by Salesforce Bulk API". Don't blame CDAP blame Salesforce.